lib/devicetrust/native/device_windows.go

```diff
@@ -0,0 +1,489 @@
+		return nil, trace.ConvertSystemError(err)
+	}
+
+	ak, err := tpm.LoadAK(ref)
+	if err != nil {
+		return nil, trace.Wrap(err, "loading ak into tpm")
+	}
+
+	return ak, nil
+}
+
+func createAndSaveAK(
+	tpm *attest.TPM,
+	persistencePath string,
+) (*attest.AK, error) {
+	ak, err := tpm.NewAK(nil)
+	if err != nil {
+		return nil, trace.Wrap(err, "creating ak")
+	}
+
+	// Write it to the well-known location on disk
+	ref, err := ak.Marshal()
+	if err != nil {
+		return nil, trace.Wrap(err, "marshalling ak")
+	}
+	err = os.WriteFile(persistencePath, ref, 600)
+	if err != nil {
+		return nil, trace.Wrap(err, "writing ak to disk")
+	}
+
+	return ak, nil
+}
+
+func enrollDeviceInit() (*devicepb.EnrollDeviceInit, error) {
+	akPath, err := setupDeviceStateDir(os.UserHomeDir)
+	if err != nil {
+		return nil, trace.Wrap(err, "setting up device state directory")
+	}
+
+	tpm, err := openTPM()
+	if err != nil {
+		return nil, trace.Wrap(err, "opening tpm")
+	}
+	defer tpm.Close()
+
+	// Try to load an existing AK in the case of re-enrollment, but, if the
+	// AK does not exist, create one and persist it.
+	ak, err := loadAK(tpm, akPath)
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			return nil, trace.Wrap(err, "loading ak")
+		}
+		log.Debug("No existing AK was found on disk, an AK will be created.")
+		ak, err = createAndSaveAK(tpm, akPath)
+		if err != nil {
+			return nil, trace.Wrap(err, "creating ak")
+		}
+	} else {
+		log.Debug("Existing AK was found on disk, it will be reused.")
+	}
+	defer ak.Close(tpm)
+
+	deviceData, err := collectDeviceData()
+	if err != nil {
+		return nil, trace.Wrap(err, "collecting device data")
+	}
+
+	marshaledEK, err := getMarshaledEK(tpm)
+	if err != nil {
+		return nil, trace.Wrap(err, "marshalling ek")
+	}
+
+	credentialID, err := credentialIDFromAK(ak)
+	if err != nil {
+		return nil, trace.Wrap(err, "determining credential id")
+	}
+
+	return &devicepb.EnrollDeviceInit{
+		CredentialId: credentialID,
+		DeviceData:   deviceData,
+		Tpm: &devicepb.TPMEnrollPayload{
+			Ek: &devicepb.TPMEnrollPayload_EkKey{
+				EkKey: marshaledEK,
+			},
+			AttestationParameters: devicetrust.AttestationParametersToProto(
+				ak.AttestationParameters(),
+			),
+		},
+	}, nil
+}
+
+// credentialIDFromAK produces a deterministic short-ish unique-ish human
+// readable-ish string identifier for a given AK. This can then be used as a
+// reference for this AK in the backend.
+//
+// To produce this, we perform a SHA256 hash over the constituent fields of
+// the AKs public key and then base64 encode it to produce a human-readable
+// string. This is similar to how SSH fingerprinting of public keys work.
+func credentialIDFromAK(ak *attest.AK) (string, error) {
+	akPub, err := attest.ParseAKPublic(
+		attest.TPMVersion20,
+		ak.AttestationParameters().Public,
+	)
+	if err != nil {
+		return "", trace.Wrap(err, "parsing ak public")
+	}
+	publicKey := akPub.Public
+	switch publicKey := publicKey.(type) {
+	case *rsa.PublicKey:
```

Comment:
Should we code for ecdsa too, or would that never happen?

Reply:
`go-attestestation` seems to only internally support `*rsa.PublicKey` - so won't return ecdsa. We may want to "fix" `go-attestestation`  at some-point in the future to support ecdsa though.

Reply:
I've added some comments to this affect in the code

Reply:
I think we can write this snippet with reasonable confidence, even if we can't trigger it, but fair enough.